icefoganalytics/travel-authorization
# Fix Travel Desk Enough To See Flight Options

Relates to:

- https://github.com/icefoganalytics/travel-authorization/issues/231
- [Wireframes, Travel Desk, 2024-04-17.pdf](https://github.com/icefoganalytics/travel-authorization/blob/main/_Design/Wireframes%2C%20Travel%20Desk%2C%202024-04-17.pdf) page 9
- https://github.com/icefoganalytics/travel-authorization/issues/230

# Context

Create the Rank travel desk options screens for Traveler (https://travel-auth-dev.ynet.gov.yk.ca/my-travel-requests/227/request/rank-options).
Pull the grouped segments from travel desk forms across

![Image](https://github.com/user-attachments/assets/b22002ad-43e7-415d-b59e-32a0e06e9f62)

# Implementation

1. Rebuilt ProcessTravelDeskRequest (travel desk edit modal) as a page to avoid _some_ of the modal hell, and vastly simplify the travel desk table implementation logic.
2. Fix serialization issues in legacy traveldesk-router.
3. Fix render of flight request options on travel desk edit page.
4. Simplify travel desk "confirm booking" modal via dedicated back-end endpoint.
5. Improve styling of traveler details card.
6. Make validation of Purpose form card lazy, so it doesn't show as invalid after first creating a travel authorization.
7. Fix dashboard create travel request button.

## Concerns

1. The current Travel Desk components are using some very non-traditional and hard to comprehend patterns. Half of the code does not work, or was only half implemented. It will be faster to rebuild these components before trying to solve https://github.com/icefoganalytics/travel-authorization/issues/230, than trying to work with the existing code.
2. The "flight options" Preference select was never implemented in the travel desk edit modal (now page). It still isn't.

# Screenshots

Travel Desk with sizing issues (old)
http://localhost:8080/travel-desk
![image](https://github.com/user-attachments/assets/a6bb4bee-67bc-4883-bba6-921a3c33a021)

Travel Desk with sizing issues fixed (new)
http://localhost:8080/travel-desk
![image](https://github.com/user-attachments/assets/5d927c6e-8bb7-4a76-94b9-b25f43f9c92b)

Travel Desk Edit Modal with crash on flight option render (old)
http://localhost:8080/travel-desk
![image](https://github.com/user-attachments/assets/c88002d0-471e-4698-914a-1cc66ae29e70)

Travel Desk Edit as Page with flight option render fixed (new)
http://localhost:8080/travel-desk/41/edit
![image](https://github.com/user-attachments/assets/3f8b76cc-1491-4d56-b86c-2763cd6f441a)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
3. Boot the app via `dev up`
4. Log in to the app at http://localhost:8080
5. From the dashboard, scroll to the bottom and click "New Travel Request" to create a travel authorization.
6. Note that the Purpose form card no longer shows as immediately invalid.
7. Fill in the Purpose and Details cards, and the generate an estimate.
8. Submit the travel request to an account you controller that has access to the travel desk.
9. Log in as the second account.
10. From the top navigation, click "Manager View", and approve the request.
11. Log back in as the original account.
12. From the top navigation, click "My Travel Requests".
13. Go to the last page of the table, and click on the "Submit Travel Desk Request" button.
14. Fill out the Traveler Details and then submit the request.
15. Log in as the second account.
16. From the top navigation, click "Travel Desk".
17. Note that the table now fits in the page better, and renders with less lag.
18. Click on the "Edit" button for the last request.
19. Note that the modal is now a page with simplified styling.
20. Find the "Travel Port Text" button and click on it to open the modal.
21. Paste in this text
    ```
    Passenger:
    Pasenger details here...

    Flights:
    Air Canada AC303
    Departure: 13 Jan 09:05 Pierre Elliott Trudeau Intl Arpt (YUL) Terminal:
    Arrival: 13 Jan 11:53 Vancouver Intl Arpt (YVR) Terminal: M
    Duration: 5 Hour(s) 48 Minutes
    Status: Confirmed
    Class: Q
    Air Canada AC304
    Departure: 13 Jan 09:05 Pierre Elliott Trudeau Intl Arpt (YUL) Terminal:
    Arrival: 13 Jan 11:53 Vancouver Intl Arpt (YVR) Terminal: M
    Duration: 5 Hour(s) 48 Minutes
    Status: Confirmed
    Class: Q

    Hotels:
    Hotel details here...

    Cars:
    Car rental details here...
    ```
22. Click "Clean and speparate options".
23. Select both segments and click "Group Selected".
24. Add a cost and select a "leg", then click "Save Groupings".
25. Scroll to the "Flight Request" table and expand the row via the "down arrow" button at the start of the row.
